File: app/src/main/java/com/ravimishra/tradzhub/Activity/ProductActivity.java
package com.ravimishra.tradzhub.Activity;

import android.content.Intent;
import android.content.SharedPreferences;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.widget.NestedScrollView;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.ravimishra.tradzhub.Model.CategoryModel;
import com.ravimishra.tradzhub.Model.TradzHubProductModel;
import com.ravimishra.tradzhub.R;
import com.ravimishra.tradzhub.Utils.Constants;

public class ProductActivity extends AppCompatActivity implements View.OnClickListener {
    public static final String DATABASE_NAME = "cartdatabase";
    SQLiteDatabase mDatabase;

    private TextView[] bottomBars;
    private ImageView backImage;
    private LinearLayout Layout_bars, llAddToCart, llBuyNow;
    private MyViewPagerAdapter myvpAdapter;
    private Drawable banner2, banner3, banner4;
    private TextView tvProductName, tvProductPrice, tvProductDescription, tvInStock, tvOutStock, tvOrignalPrice, tvOffPrice, tvShippingCharges;
    private NestedScrollView scrollView;
    private LinearLayout bottomLL;
    private ProgressBar progressBar;
    private TradzHubProductModel.ResponseData responseData;
    private LinearLayout tvBuyNow;
    private CategoryModel.ResponseData catResponseData;

    private int value = 0;
    private ViewPager.OnPageChangeListener viewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {

        @Override
        public void onPageSelected(int position) {
            ColoredBars(position);
        }

        @Override
        public void onPageScrolled(int arg0, float arg1, int arg2) {
        }

        @Override
        public void onPageScrollStateChanged(int arg0) {

        }
    };
    private ViewPager viewPager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_product);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        viewPager = findViewById(R.id.viewPager);
        Layout_bars = findViewById(R.id.layoutBars);
        tvBuyNow = findViewById(R.id.tvBuyNow);
        tvOrignalPrice = findViewById(R.id.productOrignalPrice);

        tvOffPrice = findViewById(R.id.tvOffPrice);
        tvInStock = findViewById(R.id.tvInstock);
        tvOutStock = findViewById(R.id.tvOutstock);
        tvShippingCharges = findViewById(R.id.tvShippingCharges);
        tvOffPrice = findViewById(R.id.tvOffPrice);
        llBuyNow = findViewById(R.id.llBuyNow);
        llAddToCart = findViewById(R.id.llAddToCart);
        backImage = findViewById(R.id.back);
        llBuyNow.setOnClickListener(this);
        llAddToCart.setOnClickListener(this);
        backImage.setOnClickListener(this);
        tvBuyNow.setOnClickListener(this);

        bottomLL = findViewById(R.id.bottomLL);
        progressBar = findViewById(R.id.progressbar);
        scrollView = findViewById(R.id.scrollView);

        tvProductName = findViewById(R.id.productName);
        tvProductPrice = findViewById(R.id.productPrice);
        tvProductDescription = findViewById(R.id.productDescription);
        Bundle extras = getIntent().getExtras();
        value = extras.getInt("FROM");
        if (value == 1) {
            setData();
        }

        myvpAdapter = new MyViewPagerAdapter();
        viewPager.setAdapter(myvpAdapter);
        viewPager.addOnPageChangeListener(viewPagerPageChangeListener);
        ColoredBars(1);
        showLoader();

    }

    private void ColoredBars(int thisScreen) {
        int[] colorsInactive = getResources().getIntArray(R.array.dot_on_page_not_active);
        int[] colorsActive = getResources().getIntArray(R.array.dot_on_page_active);
        bottomBars = new TextView[3];

        Layout_bars.removeAllViews();
        for (int i = 0; i < bottomBars.length; i++) {
            bottomBars[i] = new TextView(this);
            bottomBars[i].setTextSize(50);
            bottomBars[i].setText(Html.fromHtml("&#8226;"));
            Layout_bars.addView(bottomBars[i]);
            bottomBars[i].setTextColor(colorsInactive[thisScreen]);
        }
        if (bottomBars.length > 0)
            bottomBars[thisScreen].setTextColor(colorsActive[thisScreen]);
    }

    private void showLoader() {
        progressBar.setVisibility(View.VISIBLE);
        scrollView.setVisibility(View.GONE);
        bottomLL.setVisibility(View.GONE);
        Thread myThread = new Thread() {
            @Override
            public void run() {
                try {
                    sleep(800);
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            progressBar.setVisibility(View.GONE);
                            scrollView.setVisibility(View.VISIBLE);
                            bottomLL.setVisibility(View.VISIBLE);
                        }
                    });

                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        };
        myThread.start();
    }

    private void setData() {
        Intent i = getIntent();
        responseData = (TradzHubProductModel.ResponseData) i.getSerializableExtra("PRODUCT");
        int stockValue = Integer.parseInt(responseData.currentStock);
        float purchasedValue = Float.parseFloat(responseData.purchasePrice);
        float saleValue = Float.parseFloat(responseData.salePrice);
        float percentOff = (1 - (purchasedValue / saleValue)) * 100;
        int percentAbsolutePrice = (int) percentOff;

        tvProductName.setText(responseData.title);
        tvProductPrice.setText("$ " + responseData.purchasePrice);
        tvProductDescription.setText(responseData.description);
        tvOrignalPrice.setText("$ " + responseData.salePrice);
        tvOffPrice.setText(percentAbsolutePrice + "% off");
        tvShippingCharges.setText("$ " + responseData.shippingCost);


        if (stockValue <= 1) {
            tvInStock.setVisibility(View.GONE);
            tvOutStock.setVisibility(View.VISIBLE);
        } else {
            tvOutStock.setVisibility(View.GONE);
            tvInStock.setVisibility(View.VISIBLE);
        }
    }

    private int getItem(int i) {
        return viewPager.getCurrentItem() + i;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.back:
                finish();
                break;
            case R.id.llAddToCart:
                addItemToCart();
                break;

        }
    }

    private void addItemToCart() {
        String cartItem;
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        cartItem = preferences.getString(Constants.SHARED_CART_ITEM, "");
        SharedPreferences.Editor editor = preferences.edit();
        cartItem = cartItem + "," + responseData.productID;
        editor.putString(Constants.SHARED_CART_ITEM, cartItem);
        editor.apply();


    }

    public class MyViewPagerAdapter extends PagerAdapter {
        private LayoutInflater inflater;

        public MyViewPagerAdapter() {
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            inflater = (LayoutInflater) getSystemService(ProductActivity.LAYOUT_INFLATER_SERVICE);
            View view = inflater.inflate(R.layout.pager_item, container, false);
            ImageView img = view.findViewById(R.id.productImage);
            if (value == 1) {
                RequestOptions options = new RequestOptions()
                        .centerCrop()
                        .placeholder(R.drawable.place_holder_image)
                        .error(R.drawable.place_holder_image);
                Glide.with(ProductActivity.this).load(responseData.productImage).apply(options).into(img);
            }
            //  img.setImageDrawable(topBannerList.get(position));
            container.addView(view);
            return view;
        }

        @Override
        public int getCount() {
            return 3;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            View v = (View) object;
            container.removeView(v);
        }

        @Override
        public boolean isViewFromObject(View v, Object object) {
            return v == object;
        }
    }
}

File: app/src/main/java/com/ravimishra/tradzhub/Activity/ProductActivity.kt
package com.ravimishra.tradzhub.Activity

import android.content.Intent
import android.database.sqlite.SQLiteDatabase
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.text.Html
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.widget.NestedScrollView
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.ravimishra.tradzhub.Model.Category
import com.ravimishra.tradzhub.Model.CategoryModel
import com.ravimishra.tradzhub.Model.Product
import com.ravimishra.tradzhub.R
import kotlinx.android.synthetic.main.content_product.*
import java.util.*

class ProductActivity : AppCompatActivity(), View.OnClickListener {
    var mDatabase: SQLiteDatabase? = null
    private var bottomBars: Array<TextView?>? = null

    private var Layout_bars: LinearLayout? = null

    // private var llAddToCart: LinearLayout? = null
    private var myvpAdapter: MyViewPagerAdapter? = null
    private val banner2: Drawable? = null
    private val banner3: Drawable? = null
    private val banner4: Drawable? = null
    private var tvProductName: TextView? = null
    private var tvProductPrice: TextView? = null
    private var tvProductDescription: TextView? = null
    private var tvInStock: TextView? = null
    private var tvOutStock: TextView? = null
    private var tvOrignalPrice: TextView? = null
    private var tvOffPrice: TextView? = null
    private var tvShippingCharges: TextView? = null
    private var scrollView: NestedScrollView? = null
    private var bottomLL: LinearLayout? = null
    private var progressBar: ProgressBar? = null
    private var productModel: Product? = null
    private val catResponseData: CategoryModel.ResponseData? = null
    private var value = 0
    private lateinit var stockArray: Array<Boolean>

    private val viewPagerPageChangeListener: OnPageChangeListener = object : OnPageChangeListener {
        override fun onPageSelected(position: Int) {
            ColoredBars(position)
        }

        override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {}
        override fun onPageScrollStateChanged(arg0: Int) {}
    }

    //    private var viewPager: ViewPager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_product)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        Layout_bars = findViewById(R.id.layoutBars)
        tvOrignalPrice = findViewById(R.id.productOrignalPrice)
        tvOffPrice = findViewById(R.id.tvOffPrice)
        tvInStock = findViewById(R.id.tvInstock)
        tvOutStock = findViewById(R.id.tvOutstock)
        tvShippingCharges = findViewById(R.id.tvShippingCharges)
        tvOffPrice = findViewById(R.id.tvOffPrice)
        bottomLL = findViewById(R.id.bottomLL)
        progressBar = findViewById(R.id.progressbar)
        scrollView = findViewById(R.id.scrollView)
        tvProductName = findViewById(R.id.productName)
        tvProductPrice = findViewById(R.id.productPrice)
        tvProductDescription = findViewById(R.id.productDescription)
        llAddToCart.setOnClickListener(this)
        val extras = intent.extras
        value = extras!!.getInt("FROM")

        setData()

        myvpAdapter = MyViewPagerAdapter()
        viewPager.adapter = myvpAdapter
        viewPager.addOnPageChangeListener(viewPagerPageChangeListener)
        ColoredBars(1)
        showLoader()
    }

    private fun ColoredBars(thisScreen: Int) {
        val colorsInactive = resources.getIntArray(R.array.dot_on_page_not_active)
        val colorsActive = resources.getIntArray(R.array.dot_on_page_active)
        bottomBars = arrayOfNulls(3)
        Layout_bars!!.removeAllViews()
        for (i in bottomBars!!.indices) {
            bottomBars!![i] = TextView(this)
            bottomBars!![i]?.textSize = 40f
            bottomBars!![i]?.text = Html.fromHtml("&#8226;")
            Layout_bars!!.addView(bottomBars!![i])
            bottomBars!![i]?.setTextColor(colorsInactive[thisScreen])
        }
        if (bottomBars!!.isNotEmpty()) bottomBars!![thisScreen]?.setTextColor(colorsActive[thisScreen])
    }

    private fun showLoader() {
        progressBar!!.visibility = View.VISIBLE
        scrollView!!.visibility = View.GONE
        bottomLL!!.visibility = View.GONE
        val myThread: Thread = object : Thread() {
            override fun run() {
                try {
                    sleep(800)
                    runOnUiThread {
                        progressBar!!.visibility = View.GONE
                        scrollView!!.visibility = View.VISIBLE
                        bottomLL!!.visibility = View.VISIBLE
                    }
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
        myThread.start()
    }

    private fun setData() {
        val intent = intent
        productModel = intent.getSerializableExtra("PRODUCT") as Product

        var orignalPrice = productModel?.price
        var discountPercent = productModel?.discount
        var discountedPrice = orignalPrice!! * (1 - discountPercent!! / 100)
        tvProductName!!.text = productModel!!.name
        tvProductPrice!!.text = "₹ $discountedPrice"
        tvOrignalPrice!!.text = "$ $orignalPrice"
        tvOffPrice?.text = "$discountPercent% off"
//        if (stockArray.random()) {
//            tvInStock!!.visibility = View.GONE
//            tvOutStock!!.visibility = View.VISIBLE
//        } else {
        tvOutStock!!.visibility = View.GONE
        tvInStock!!.visibility = View.VISIBLE
        // }
    }

    private fun getItem(i: Int): Int {
        return viewPager!!.currentItem + i
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.back -> finish()
            R.id.llAddToCart -> addItemToCart()
            R.id.cart -> {
                val intent = Intent(this@ProductActivity, CartActivity::class.java)
                startActivity(intent)
            }
            R.id.ivWishlist -> {
                val intent1 = Intent(this@ProductActivity, WishlistActivtry::class.java)
                startActivity(intent1)
            }
        }
    }

    private fun addItemToCart() {
        val intent = intent
        productModel = intent.getSerializableExtra("PRODUCT") as Product
        val category = productModel?.category
        if (category != null) {
            val database = FirebaseDatabase.getInstance("https://tradzhub-58133-default-rtdb.firebaseio.com")
            val myRef = database.getReference("product")
            myRef.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    for (ds in dataSnapshot.children) {

                        for (value in ds.children) {
                            val productHashMap: HashMap<String, Any> = value.value as HashMap<String, Any>
                            for (product in productHashMap.keys) {
                                if (productModel!!.id == productHashMap[product]) {
                                    val ref = value.ref

                                    ref.child("cart").setValue("1")
                                    Toast.makeText(applicationContext, "Item added to cart", Toast.LENGTH_SHORT)
                                }
                            }
                        }
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    // Failed to read value
                    Log.w("Tag", "Failed to read value.", error.toException())
                }
            })
        }
    }
    private fun addItemTOWishList(){
        val intent = intent
        productModel = intent.getSerializableExtra("PRODUCT") as Product
        val category = productModel?.category
        if (category != null) {
            val database = FirebaseDatabase.getInstance("https://tradzhub-58133-default-rtdb.firebaseio.com")
            val myRef = database.getReference("product")
            myRef.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    for (ds in dataSnapshot.children) {

                        for (value in ds.children) {
                            val productHashMap: HashMap<String, Any> = value.value as HashMap<String, Any>
                            for (product in productHashMap.keys) {
                                if (productModel!!.id == productHashMap[product]) {
                                    val ref = value.ref

                                    ref.child("cart").setValue("1")
                                    Toast.makeText(applicationContext, "Item added to cart", Toast.LENGTH_SHORT)
                                }
                            }
                        }
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                    // Failed to read value
                    Log.w("Tag", "Failed to read value.", error.toException())
                }
            })
        }
    }

    inner class MyViewPagerAdapter : PagerAdapter() {
        private var inflater: LayoutInflater? = null
        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            inflater = getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val view = inflater!!.inflate(R.layout.pager_item, container, false)
            val img = view.findViewById<ImageView>(R.id.productImage)
            //            if (value == 1) {
            val options = RequestOptions()
                    .centerCrop()
                    .placeholder(R.drawable.place_holder_image)
                    .error(R.drawable.place_holder_image)
            Glide.with(this@ProductActivity).load(productModel?.imgUrl).apply(options).into(img)
            //            }
            //  img.setImageDrawable(topBannerList.get(position));
            container.addView(view)
            return view
        }

        override fun getCount(): Int {
            return 3
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            val v = `object` as View
            container.removeView(v)
        }

        override fun isViewFromObject(v: View, `object`: Any): Boolean {
            return v === `object`
        }
    }

    companion object {
        const val DATABASE_NAME = "cartdatabase"
    }
}